### PR TITLE
QF-1551: Implement storage limit enforcement

### PR DESCRIPTION
**Pull request by olivierdalang**
(based on QF-1555 that must be merged/reviewed first)

Questions/comments:

- where exactly do we want to enforce this ? currently, it's done at the API level, but we may want to move this closer to the model if we want to reuse that same logic also from other places (admin, web frontend...). Same question for both file uploads and project reparenting
- currently, no caching is done about storage size, meaning that each time, each time the quota is checked (e.g. each upload), requests are made for **all versions of all files of all projects of the user**. I think we need to cache this at the project level. But maybe this already justifies investigating a bit more the idea of using django-storages and keep track of the files in the database ?

### CODE REVIEW

**Comment by duke-nyuki** [comment]
Task linked: [QF-1555 Design the database schema for subscriptions](https://app.clickup.com/t/2192114/QF-1555)

**Comment by duke-nyuki** [comment]
Task linked: [QF-1551 Implement the limitations for storage size limit](https://app.clickup.com/t/2192114/QF-1551)

File: docker-app/qfieldcloud/core/utils2/storage.py
Comment (suricactus) [inline docker-app/qfieldcloud/core/utils2/storage.py:197]:
```suggestion
def file_size_bytes(file: IO) -> int:
```

Comment (suricactus) [reply docker-app/qfieldcloud/core/utils2/storage.py:197]: I think this is the best way to do it in python...

File: docker-app/qfieldcloud/core/views/files_views.py
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/files_views.py:162]:
```suggestion
        file_size_mb = request_file.size / 1024 / 1024
```
[Don't you have something like this?](https://docs.djangoproject.com/en/4.0/ref/files/uploads/#django.core.files.uploadedfile.UploadedFile.size)

**Comment by suricactus** [comment]
> * where exactly do we want to enforce this ? currently, it's done at the API level, but we may want to move this closer to the model if we want to reuse that same logic also from other places (admin, web frontend...). Same question for both file uploads and project reparenting

Closer to the model, the better. I think the admin iface is the only exception to the enforcement of these checks.

> * currently, no caching is done about storage size, meaning that each time, each time the quota is checked (e.g. each upload), requests are made for **all versions of all files of all projects of the user**. I think we need to cache this at the project level. But maybe this already justifies investigating a bit more the idea of using django-storages and keep track of the files in the database ?

My two main concerns for "storing files info in the database" are:
- how are we sure we keep db and s3 in sync? 140% need for monitoring from day 0!
- how are we sure our transaction is finished: both s3 stored the file and the db saved a new record for the file?

If these two are solved, this would open a great area of optimizations. Give it a thought.

If not - project level cache is reasonable.

File: docker-app/qfieldcloud/core/views/files_views.py
Comment (olivierdalang) [reply docker-app/qfieldcloud/core/views/files_views.py]: Indeed (got confused because that doesn't exist for io.BytesIO, but here we deal with a django UploadedFile)

File: docker-app/qfieldcloud/core/models.py
Comment (olivierdalang) [reply docker-app/qfieldcloud/core/models.py:427]: I think there's some kind of convention that says that property shouldn't be used for heavy lifting methods (as otherwise you can easily inadvertently make many queries), but will change once cache is in place

File: docker-app/qfieldcloud/core/migrations/0054_remove_project_files_count.py
Comment (suricactus) [inline docker-app/qfieldcloud/core/migrations/0054_remove_project_files_count.py:1]: Can we squash this migration with 0052?

File: docker-app/qfieldcloud/core/views/files_views.py
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/files_views.py:163]:
```suggestion
        quota_left_mb = project.owner.useraccount.storage_quota_left_mb
```

File: docker-app/qfieldcloud/core/views/projects_views.py
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/projects_views.py:134]: I tried to thing if we can check this before opening the transaction, but failed to find a solution. Reminder just in case you have an idea to offload postgres bit.

File: docker-app/qfieldcloud/subscription/tests/test_storage_quota.py
Comment (suricactus) [inline docker-app/qfieldcloud/subscription/tests/test_storage_quota.py:40]:
```suggestion
        account_type_1mb = AccountType.objects.create(code="account_type_1mb", storage_mb=1)
```

File: docker-app/qfieldcloud/subscription/tests/test_storage_quota.py
Comment (suricactus) [inline docker-app/qfieldcloud/subscription/tests/test_storage_quota.py:46]:
```suggestion
        extra_2mb = ExtraPackageTypeStorage.objects.create(code="extra_2mb", megabytes=2)
```
Sorry, these cryptic names are already confusing for me.

File: docker-app/qfieldcloud/subscription/tests/test_storage_quota.py
Comment (suricactus) [inline docker-app/qfieldcloud/subscription/tests/test_storage_quota.py:104]:
```suggestion
        account_type_1mb = AccountType.objects.create(code="account_type_1mb", storage_mb=1)
```

File: docker-app/qfieldcloud/subscription/tests/test_storage_quota.py
Comment (suricactus) [inline docker-app/qfieldcloud/subscription/tests/test_storage_quota.py:129]:
```suggestion
        extra_1mb = ExtraPackageTypeStorage.objects.create(
            code="extra_1mb", display_name="onemegextra", megabytes=1
        )
```

File: docker-app/qfieldcloud/core/models.py
Comment (suricactus) [inline docker-app/qfieldcloud/core/models.py:950]: Why did you convert to a list?

**Comment by suricactus** [comment]
Thanks! Except a few typing and naming details, I believe this matches the requirements.

I also added a few extra questions to explain the logic behind some decisions, but should not be considered blocking.

File: docker-app/qfieldcloud/core/models.py
Comment (olivierdalang) [reply docker-app/qfieldcloud/core/models.py:905]: ok

File: docker-app/qfieldcloud/core/migrations/0054_remove_project_files_count.py
Comment (olivierdalang) [reply docker-app/qfieldcloud/core/migrations/0054_remove_project_files_count.py:1]: ok

File: docker-app/qfieldcloud/core/views/projects_views.py
Comment (olivierdalang) [reply docker-app/qfieldcloud/core/views/projects_views.py:134]: We could compare username in the json to the username of the instance, but not sure this is robust before validation (e.g. it could mix upper/lowercases, so that old_ownername != new_owndername, even if they represent the same object). And if we were to validate it before, we'd hit the DB anyway.

In any case, this doesn't happen too frequently, IMO not making a difference for the DB ? But I don't mind too much, I can change to the above idea if you prefer avoiding transaction (indeed a bit counterinutitive to do a change and rollback afterwards).

I switched to decorator syntax though.

File: docker-app/qfieldcloud/core/models.py
Comment (suricactus) [reply docker-app/qfieldcloud/core/models.py:950]: Yes, clear! Thanks! Now I saw your commit message too.

File: docker-app/qfieldcloud/core/models.py
Comment (suricactus) [reply docker-app/qfieldcloud/core/models.py:1010]: clear. thanks!

**Comment by suricactus** [comment]
Second review from my side and seems this PR is ready. A few steps before merging is to fix the test and squash some extra commits, but seems we are ready to go.

**Review by suricactus — APPROVED** [review]

File: docker-app/qfieldcloud/core/views/projects_views.py
Comment (suricactus) [reply docker-app/qfieldcloud/core/views/projects_views.py]: It is good, thanks!

**Comment by olivierdalang** [comment]
Ok I think we can squash this in just one commit ?

About 
> Closer to the model, the better. I think the admin iface is the only exception to the enforcement of these checks.

I didn't really change that, so it must still be done separately in each view. IMO it would be a good move, but only makes sense with a larger refactoring (moving all file related stuff to the Projects model, and letting that model take responsibility for adding/deleting files).

**Pull request merged**
